logins-sql/ffi/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+                access_token: c_str_to_str(access_token).into(),
+                tokenserver_url: parse_url(c_str_to_str(tokenserver_url))?,
+            },
+            &sync15_adapter::KeyBundle::from_ksync_base64(
+                c_str_to_str(sync_key).into()
+            )?
+        )
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn sync15_passwords_touch(
+    state: *const PasswordEngine,
+    id: *const c_char,
+    error: *mut ExternError
+) {
+    trace!("sync15_passwords_touch");
+    with_translated_void_result(error, || {
+        assert!(!state.is_null(), "Null state passed to sync15_passwords_touch");
+        let state = &*state;
+        state.touch(c_str_to_str(id))
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn sync15_passwords_delete(
+    state: *const PasswordEngine,
+    id: *const c_char,
+    error: *mut ExternError
+) -> bool {
+    trace!("sync15_passwords_delete");
+    with_translated_value_result(error, || {
+        assert!(!state.is_null(), "Null state passed to sync15_passwords_delete");
+        let state = &*state;
+        state.delete(c_str_to_str(id))
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn sync15_passwords_wipe(
+    state: *const PasswordEngine,
+    error: *mut ExternError
+) {
+    trace!("sync15_passwords_wipe");
+    with_translated_void_result(error, || {
+        assert!(!state.is_null(), "Null state passed to sync15_passwords_wipe");
+        let state = &*state;
+        state.wipe()
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn sync15_passwords_reset(
+    state: *const PasswordEngine,
+    error: *mut ExternError
+) {
+    trace!("sync15_passwords_reset");
+    with_translated_void_result(error, || {
+        assert!(!state.is_null(), "Null state passed to sync15_passwords_reset");
+        let state = &*state;
+        state.reset()
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn sync15_passwords_get_all(
+    state: *const PasswordEngine,
+    error: *mut ExternError
+) -> *mut c_char {
+    trace!("sync15_passwords_get_all");
+    with_translated_string_result(error, || {
+        assert!(!state.is_null(), "Null state passed to sync15_passwords_get_all");
+        let state = &*state;
+        let all_passwords = state.list()?;
+        let result = serde_json::to_string(&all_passwords)?;
+        Ok(result)
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn sync15_passwords_get_by_id(
+    state: *const PasswordEngine,
+    id: *const c_char,
+    error: *mut ExternError
+) -> *mut c_char {
+    trace!("sync15_passwords_get_by_id");
+    with_translated_opt_string_result(error, || {
+        assert!(!state.is_null(), "Null state passed to sync15_passwords_get_by_id");
+        let state = &*state;
+        if let Some(password) = state.get(c_str_to_str(id))? {
+            Ok(Some(serde_json::to_string(&password)?))
+        } else {
+            Ok(None)
+        }
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn sync15_passwords_add(
```

Comment:
If I did this I'd do it by exposing PasswordAction as a C struct to kotlin directly. At the moment, the FFI code is subtle and error prone to write and calling it from kotlin is more or less the same. The fewer functions we can have exposed over it the better.

That said, I'm not sure I see the benefit here. Exposing a `#[repr(C)] struct LoginC` or whatever over the FFI that gets returned and taken by value wouldn't really bother me, it's just more code and, again, I'd like to keep the FFI small since it's currently tricky. I think it would be less code than what the PasswordAction system would look like, even if it were exposed directly.

Reply:
I generally like moving toward a builder pattern.  I've been thinking about cross-product testing, and I'm led to a description of incoming changes as Sync 1.5 records on the wire, and a description of local changes as operational transforms.  Such transforms are basically the builder pattern.  So if we baked that notion of operational transform into the implementation at a foundational level, we'd be saved from deducing them after the fact (which is what we do right now to merge smoothly).

Evolving this definitely doesn't block landing -- definitely not! -- but it's something that we should be thinking about and driving, gently, towards.